src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2747,13 +2811,580 @@ def query(self, filters=None, options=None):
+            return ds['name'], {'encryption_key': key, **ds} if options.get('full_output') else key
+
+        def check_key(ds):
+            return options.get('all') or (ds['key_loaded'] and key_loaded) or (not ds['key_loaded'] and not key_loaded)
+
+        return dict(map(
+            normalize,
+            filter(
+                lambda d: d['name'] == d['encryption_root'] and d['encrypted'] and recursive_op(
+                    d['name'], name
+                ) and check_key(d),
+                self.query()
+            )
+        ))
+
+    @periodic(86400)
+    @private
+    @job(lock=lambda args: f'sync_encrypted_pool_dataset_keys_{args}')
+    def sync_db_keys(self, job, filters=None):
+        db_datasets = self.encrypted_roots_query_db(filters=filters, decrypt=True)
+        to_remove = []
+        check_key_job = self.middleware.call_sync('zfs.dataset.bulk_process', 'check_key', [
+            (name, {'key': db_datasets[name]}) for name in db_datasets
+        ])
+        check_key_job.wait_sync()
+        if check_key_job.error:
+            self.middleware.logger.error(f'Failed to sync database keys: {check_key_job.error}')
+
+        for dataset, status in zip(db_datasets, check_key_job.result):
+            if not status['result'] or status['error']:
+                to_remove.append(dataset)
+
+        self.middleware.call_sync('datastore.delete', self.dataset_store, [['name', 'in', to_remove]])
+
+    @accepts(Str('id'))
+    @job(lock='dataset_export_keys', pipes=['output'])
+    def export_keys(self, job, id):
+        """
+        Export keys for `id` and its children which are stored in the system. The exported file is a compressed
+        tarfile containing files named after the encrypted dataset and containing the keys for them.
+
+        Please refer to websocket documentation for downloading the file.
+        """
+        self.middleware.call_sync('pool.dataset._get_instance', id)
+        sync_job = self.middleware.call_sync(
+            'pool.dataset.sync_db_keys', ['OR', [['name', '=', id], ['name', '^', f'{id}/']]]
+        )
+        sync_job.wait_sync()
+
+        datasets = self.encrypted_roots_query_db(
+            ['OR', [['name', '=', id], ['name', '^', f'{id}/']]], decrypt=True
+        )
+        temp_dir = None
+        try:
+            temp_dir = tempfile.mkdtemp()
+            tar_path = os.path.join(temp_dir, 'keys')
+            with tarfile.open(tar_path, 'w:xz') as tar:
+                for ds in datasets:
+                    db_key = datasets[ds]
+                    key = io.BytesIO(db_key if isinstance(db_key, bytes) else db_key.encode())
+                    info = tarfile.TarInfo(name=ds)
+                    info.size = len(db_key)
+                    tar.addfile(tarinfo=info, fileobj=key)
+
+            with open(tar_path, 'rb') as f:
+                shutil.copyfileobj(f, job.pipes.output.w)
+        finally:
+            shutil.rmtree(temp_dir or '', ignore_errors=True)
```

Comment:
In case something wrong happens and a temporary directory is not created for any reason ? However an `if` would seem more reasonable here perhaps first. I’ll do that